File: 2699-count-the-number-of-fair-pairs/count-the-number-of-fair-pairs.py
class Solution:
    def countFairPairs(self, nums: List[int], lower: int, upper: int) -> int:
        nums.sort()

        def binarySearch(l, r, target):
            while l <= r:
                m = l + (r - l)//2
                if nums[m] >= target:
                    r = m - 1
                else:
                    l = m + 1
            return r
        
        result = 0
        for i in range(len(nums)):
            small = lower - nums[i]
            large =  upper + 1 - nums[i]
            result += (binarySearch(i + 1, len(nums) - 1, large) - binarySearch(i + 1, len(nums) - 1, small))
        return result

File: 2699-count-the-number-of-fair-pairs/count-the-number-of-fair-pairs.cpp
class Solution {
public:
    int binary(int l, int r, int target, vector<int>& nums){
        while (l <= r){
            int m = l + int((r - l) / 2);
            if (nums[m] >= target) r = m - 1;
            else l = m + 1;
        }
        return r;
    }

    long long countFairPairs(vector<int>& nums, int lower, int upper) {
        sort(nums.begin(), nums.end());

        long long result = 0;

        for (int i = 0; i < nums.size() - 1; i++){
            int l_target = lower - nums[i];
            int r_target = upper - nums[i] + 1;
            result += (binary(i + 1, nums.size() - 1, r_target, nums) - binary(i + 1, nums.size() - 1, l_target, nums));
        }
        return result;
    }
};